engine/table/src/main/java/io/deephaven/engine/table/impl/lang/QueryLanguageParser.java

```diff
@@ -1738,38 +1738,58 @@ public Class<?> visit(MethodCallExpr n, VisitArgs printer) {
         // evaluation code that expects singular values. This check makes sure that numba vectorized functions must be
         // used alone (or with cast only) as the entire expression.
-        n.getParentNode().ifPresent(parent -> {
-            if (parent.getClass() != CastExpr.class || parent.getParentNode().isPresent()) {
-                throw new RuntimeException("Numba vectorized function can't be used in an expression.");
-            }
-        });
 
         final QueryScope queryScope = ExecutionContext.getContext().getQueryScope();
         for (QueryScopeParam<?> param : queryScope.getParams(queryScope.getParamNames())) {
             if (param.getName().equals(n.getNameAsString())) {
-                NumbaCallableWrapper numbaCallableWrapper = (NumbaCallableWrapper) param.getValue();
-                List<Class<?>> params = numbaCallableWrapper.getParamTypes();
-                if (params.size() != expressions.length) {
-                    throw new RuntimeException("Numba vectorized function argument count mismatch: " + params.size()
-                            + " vs." + expressions.length);
-                }
-                for (int i = 0; i < expressions.length; i++) {
-                    if (!(expressions[i] instanceof NameExpr)) {
-                        throw new RuntimeException("Numba vectorized function arguments can only be columns.");
-                    }
-                    if (!isSafelyCoerceable(expressionTypes[i], params.get(i))) {
-                        throw new RuntimeException("Numba vectorized function argument type mismatch: "
-                                + expressionTypes[i].getSimpleName() + " -> " + params.get(i).getSimpleName());
-                    }
+                CallableWrapper callableWrapper = (CallableWrapper) param.getValue();
+                try {
+                    checkVectorizability(n, expressions, expressionTypes, callableWrapper);
+                } catch (RuntimeException ex) {
+                    if (callableWrapper.isVectorized())
+                        throw ex;
                 }
+                break;
+            }
+        }
+    }
+
+    private void checkVectorizability(MethodCallExpr n, Expression[] expressions, Class<?>[] expressionTypes,
+            CallableWrapper callableWrapper) {
+
+        callableWrapper.parseSignature();
+
+        n.getParentNode().ifPresent(parent -> {
+            if (parent.getClass() != CastExpr.class || parent.getParentNode().isPresent()) {
+                throw new RuntimeException("Python vectorized function can't be used in an expression.");
+            }
+            // TODO what to do with 'cast'?
+        });
+
+        List<Class<?>> params = callableWrapper.getParamTypes();
+        if (params.size() != expressions.length) {
+            throw new RuntimeException("Python vectorized function argument count mismatch: " + params.size()
+                    + " vs." + expressions.length);
+        }
```

Comment:
Functions with default parameters, zero parameters are not supported yet. Vectorized functions must have arguments and all the argument must either be column names or special variables (i,ii,k) which are the only chunkable sat this time.

Reply:
Is there a reason we can not easily support constants?

Reply:
How do we fail if there is a named argument?  Is the error message extremely clear about what happened?

Reply:
The current vectorization implementation bypasses the regular parsing-codegen-compilation flow, thus almost no ability to handle expressions including constants and evaluate them first for chunking purpose. That flow is another complex machinery that will take some sizable effort to hack. 

As for the use of named argument, the Java expression parser won't accept a syntax like f(x, y=10). So only positional arguments are possible. If a func is called without the keyword arguments, a 'argument count mismatch' expect will be thrown at formula compilation time if it is already vectorized, otherwise, a runtime exception will be thrown complaining missing a positional argument by JPY which is confusing but that's an inherent incompatibility between Java and Python, not insurmountable, but will take a good amount of hacking 

I think we need to document clearly that for Python functions to be callable in formulas for now they need to be positional parameter only.